### PR TITLE
CUDA `vector_add` sample project

### LINKED ISSUE #2159: cudax MVP Example
I'd like a minimal example of `cudax::` that illustrates the most important features:
- Allocate and initialize device memory
- Copy data to<->from host 
- Manage devices and create resources like streams and events
- Launch a kernel on a particular device/stream

I think it would be appropriate for our example to showcase a new and improved `vectorAdd` that showcases the following pieces of new machinery:
- `cuda::mr` data structures for allocation
- `cudax` device, stream, event data structures
- `cudax::launch` to launch a kernel 
- (anything else important I'm missing) 

Let's put this under the `cudax/examples` directory. In time, I think it will be a good idea for us to mirror some of the examples from the canonical [`cuda-samples`](https://github.com/NVIDIA/cuda-samples).

### PR DESCRIPTION
**Pull request by ericniebler**
## Description
This adds a sample project for cudax, initially populated with the standard CUDA `vector_add` sample. We will morph this into something beautiful using `cudax`.

closes #2159

### CODE REVIEW

File: cudax/samples/vector_add/vector_add.cu [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ * vector addition. It is the same as the sample illustrating Chapter 2
+ * of the programming guide with some additions like error checking.
+ */
+
+#include <stdio.h>
+
+// For the CUDA runtime routines (prefixed with "cuda_")
+#include <cuda_runtime.h>
+
+#include "helper_cuda.h"
```
[the remark this answers is not in the document]
Comment (ericniebler) [reply cudax/samples/vector_add/vector_add.cu:41]: Feel free to push changes directly to this pr.

File: cudax/samples/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# This will automatically clone CCCL from GitHub and make the exported cmake targets available
+CPMAddPackage(
+    NAME CCCL
+    GITHUB_REPOSITORY ${CCCL_REPOSITORY}
+    GIT_TAG ${CCCL_TAG}
+)
+
+# Default to building for the GPU on the current system
+if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
+  set(CMAKE_CUDA_ARCHITECTURES 86)
```
Comment (jrhemstad) [inline cudax/samples/CMakeLists.txt:39]: The problem isn't the architecture value, it's that the way this test is set up, it requires running on a GPU runner, but is ending up on a CPU runner. 

@alliepiper can help you get it sorted.

File: cudax/samples/vector_add/vector_add.cu [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  // Copy the host input vectors A and B in host memory to the device input
+  // vectors in device memory
+  printf("Copy input data from the host memory to the CUDA device\n");
+  d_A = h_A;
+  d_B = h_B;
+
+  // Define the kernel launch parameters
+  constexpr int threadsPerBlock = 256;
+  int blocksPerGrid             = (numElements + threadsPerBlock - 1) / threadsPerBlock;
+  auto dims = cudax::make_hierarchy(cudax::grid_dims(blocksPerGrid), cudax::block_dims<threadsPerBlock>());
```
Comment (pciolkosz) [inline cudax/samples/vector_add/vector_add.cu:178]: With #2001 merged we could use at_least(numElements) for the grid dimensions.
We could also try to come up with some shorthands for the entire hierarchy dimensions like that, they are super common

File: cudax/samples/vector_add/vector.cuh [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    ~__action()
+    {
+      printf("~action()\n");
+      if constexpr ((_Param & __param::_out) == __param::_out)
+      {
+        printf("about to synchronize the stream\n");
+        fflush(stdout);
+        __str_.wait(); // wait for the kernel to finish
```
Comment (ericniebler) [inline cudax/samples/vector_add/vector.cuh:128]: @pciolkosz @miscco for some reason this call to `stream_ref::wait` causes the process to terminate on my machine. the next line never executes, and the process exits with code `0`. no exception is thrown. can either of you tell me why?

Comment (pciolkosz) [reply cudax/samples/vector_add/vector.cuh:128]: Can you try with LIBCUDACXX_ENABLE_EXCEPTIONS defined?

File: cudax/samples/vector_add/vector_add.cu (hunk 3 above)
Comment (ericniebler) [reply cudax/samples/vector_add/vector_add.cu:96]: how about this for a shorthand:
```c++
auto dims = cudax::distribute<256>(numElements);
```

Comment (pciolkosz) [reply cudax/samples/vector_add/vector_add.cu:96]: I like that

File: cudax/samples/vector_add/vector.cuh [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        // TODO: use a memcpy async here
+        __d_ = __h_;
+      }
+      __dirty_ = false;
+    }
+  }
+
+  void sync_device_to_host(stream_ref __str, detail::__param_kind __p) const
+  {
+    if (__p != detail::__param_kind::_in)
```
Comment (pciolkosz) [inline cudax/samples/vector_add/vector.cuh:83]: Minor opinion: I think I would prefer this if to be in __action and not passed into sync_* function, that better fits the name and keeps _Kind usage within __action. Same for other direction sync

Comment (ericniebler) [reply cudax/samples/vector_add/vector.cuh:83]: Noted. `vector` is a temporary hack for the purpose of the demo. The real `cudax::vector` will be a different beast. I'll keep this in mind for the work though.

**Comment by pciolkosz** [comment]
mdspan (we use extents from there) needs c++20 on Windows, that's why CUDAX CI only has c++20 MSVC build

**Comment by ericniebler** [comment]
@miscco something seems to be going wrong with the mdspan concepts portability macros with msvc.

```c++
sccache "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\nvcc.exe" -forward-unknown-to-host-compiler -DLIBCUDACXX_ENABLE_EXCEPTIONS -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\lib\cmake\libcudacxx\..\..\..\include -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\thrust\thrust\cmake\..\.. -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\cub\cub\cmake\..\.. -isystem C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\cudax\lib\cmake\cudax\..\..\..\include -D_WINDOWS -Xcompiler="/W3 /GR /EHsc" -Xcompiler="-MDd  -Zi -Ob0 -Od /RTC1" -std=c++17 "--generate-code=arch=compute_86,code=[compute_86,sm_86]" -Xcompiler=/Zc:__cplusplus -Xcompiler=/Zc:preprocessor -MD -MT CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj -MF CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj.d -x cu -c C:\cccl\cudax\samples\vector_add\vector_add.cu -o CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj -Xcompiler=-FdCMakeFiles\vector_add.dir\,-FS
vector_add.cu
C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/default_accessor.h(75): error C4002: too many arguments for function-like macro invocation '__MDSPAN_PP_CAT_IMPL'
C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/macros.h(276): note: in expansion of macro '__MDSPAN_TEMPLATE_REQUIRES'
C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/macros.h(242): note: in expansion of macro '__MDSPAN_PP_CAT'
```

i get this with `/std:c++17 /Zc:__cplusplus /Zc:preprocessor`. see the build failure here: https://github.com/NVIDIA/cccl/actions/runs/10314338615/job/28552592563?pr=2160

**Comment by miscco** [comment]
> @miscco something seems to be going wrong with the mdspan concepts portability macros with msvc.
> 
> ```c++
> sccache "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\nvcc.exe" -forward-unknown-to-host-compiler -DLIBCUDACXX_ENABLE_EXCEPTIONS -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\lib\cmake\libcudacxx\..\..\..\include -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\thrust\thrust\cmake\..\.. -IC:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\cub\cub\cmake\..\.. -isystem C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\cudax\lib\cmake\cudax\..\..\..\include -D_WINDOWS -Xcompiler="/W3 /GR /EHsc" -Xcompiler="-MDd  -Zi -Ob0 -Od /RTC1" -std=c++17 "--generate-code=arch=compute_86,code=[compute_86,sm_86]" -Xcompiler=/Zc:__cplusplus -Xcompiler=/Zc:preprocessor -MD -MT CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj -MF CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj.d -x cu -c C:\cccl\cudax\samples\vector_add\vector_add.cu -o CMakeFiles\vector_add.dir\vector_add\vector_add.cu.obj -Xcompiler=-FdCMakeFiles\vector_add.dir\,-FS
> vector_add.cu
> C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/default_accessor.h(75): error C4002: too many arguments for function-like macro invocation '__MDSPAN_PP_CAT_IMPL'
> C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/macros.h(276): note: in expansion of macro '__MDSPAN_TEMPLATE_REQUIRES'
> C:\cccl\build\cudax-cpp20\cudax\samples\src\cudax_samples-build\_deps\cccl-src\libcudacxx\include\cuda/std/__mdspan/macros.h(242): note: in expansion of macro '__MDSPAN_PP_CAT'
> ```
> 
> i get this with `/std:c++17 /Zc:__cplusplus /Zc:preprocessor`. see the build failure here: https://github.com/NVIDIA/cccl/actions/runs/10314338615/job/28552592563?pr=2160

Yeah our `mdspan` implementation is cursed and currently requires C++20 for MSVC

I have a branch with a complete rewrite lying around, but need to implement `submdspan` anew

**Comment by ericniebler** [comment]
it doesn't seem to work with c++20 either: https://github.com/NVIDIA/cccl/actions/runs/10314858890/job/28553983052?pr=2160

halp!

**Comment by miscco** [comment]
@robertmaynard I believe we are in trouble here. The ci failures [here](https://github.com/NVIDIA/cccl/actions/runs/10318069371/job/28563704716?pr=2160) are referencing code that is not there anymore. 🙀

**Review by pciolkosz — APPROVED** [review]

**Pull request merged**